### ISSUE TITLE
ArrayIndexOutOfBoundsException when interacting (zooming/scrolling) on weekview that is showing only a single day

### ISSUE THREAD

**Issue by s3n-w6i** [SapuSeven/BetterUntis#165]
**Description**

1. Set the number of shown days to one in the settings
2. Return to the main view and zoom in or scroll down
3. The application will crash

<details>
  <summary><b>Logs</b></summary>

  ```
2020-12-09 21:00:31.448 6412-6412/com.sapuseven.untis.debug E/BetterUntis: Application crashed!
    java.lang.ArrayIndexOutOfBoundsException: length=64; index=64
        at com.sapuseven.untis.views.weekview.drawers.BackgroundGridDrawer.drawHourLines(BackgroundGridDrawer.kt:77)
        at com.sapuseven.untis.views.weekview.drawers.BackgroundGridDrawer.drawGrid(BackgroundGridDrawer.kt:42)
        at com.sapuseven.untis.views.weekview.drawers.BackgroundGridDrawer.draw(BackgroundGridDrawer.kt:17)
        at com.sapuseven.untis.views.weekview.WeekView.onDraw(WeekView.kt:310)
        at android.view.View.draw(View.java:21440)
        at android.view.View.updateDisplayListIfDirty(View.java:20317)
        at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:4372)
        at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:4345)
        at android.view.View.updateDisplayListIfDirty(View.java:20277)
        at android.view.View.draw(View.java:21172)
        at android.view.ViewGroup.drawChild(ViewGroup.java:4388)
        at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4149)
        at androidx.constraintlayout.widget.ConstraintLayout.dispatchDraw(ConstraintLayout.java:1824)
        at android.view.View.updateDisplayListIfDirty(View.java:20308)
        at android.view.View.draw(View.java:21172)
        at android.view.ViewGroup.drawChild(ViewGroup.java:4388)
        at androidx.coordinatorlayout.widget.CoordinatorLayout.drawChild(CoordinatorLayout.java:1277)
        at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4149)
        at android.view.View.updateDisplayListIfDirty(View.java:20308)
        at android.view.View.draw(View.java:21172)
        at android.view.ViewGroup.drawChild(ViewGroup.java:4388)
        at androidx.drawerlayout.widget.DrawerLayout.drawChild(DrawerLayout.java:1426)
        at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4149)
        at android.view.View.draw(View.java:21443)
        at android.view.View.updateDisplayListIfDirty(View.java:20317)
        at android.view.View.draw(View.java:21172)
        at android.view.ViewGroup.drawChild(ViewGroup.java:4388)
        at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4149)
        at android.view.View.updateDisplayListIfDirty(View.java:20308)
        at android.view.View.draw(View.java:21172)
        at android.view.ViewGroup.drawChild(ViewGroup.java:4388)
        at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4149)
        at android.view.View.updateDisplayListIfDirty(View.java:20308)
        at android.view.View.draw(View.java:21172)
        at android.view.ViewGroup.drawChild(ViewGroup.java:4388)
        at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4149)
        at android.view.View.updateDisplayListIfDirty(View.java:20308)
        at android.view.View.draw(View.java:21172)
        at android.view.ViewGroup.drawChild(ViewGroup.java:4388)
        at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4149)
        at android.view.View.updateDisplayListIfDirty(View.java:20308)
        at android.view.View.draw(View.java:21172)
        at android.view.ViewGroup.drawChild(ViewGroup.java:4388)
        at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4149)
        at android.view.View.draw(View.java:21443)
        at com.android.internal.policy.DecorView.draw(DecorView.java:806)
        at android.view.View.updateDisplayListIfDirty(View.java:20317)
        at android.view.ThreadedRenderer.updateViewTreeDisplayList(ThreadedRenderer.java:575)
        at android.view.ThreadedRenderer.updateRootDisplayList(ThreadedRenderer.java:581)
        at android.view.ThreadedRenderer.draw(ThreadedRenderer.java:654)
        at android.view.ViewRootImpl.draw(ViewRootImpl.java:3611)
        at android.view.ViewRootImpl.performDraw(ViewRootImpl.java:3419)
        at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2756)
        at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1722)
        at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:7605)
        at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1029)
        at android.view.Choreographer.doCallbacks(Choreographer.java:852)
        at android.view.Choreographer.doFrame(Choreographer.java:787)
        at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1014)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
2020-12-09 21:00:31.448 6412-6412/com.sapuseven.untis.debug E/BetterUntis:     at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7397)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:935)
```
</details>

**Additional information**

Poking around in `BackgroundGridDrawer.kt`, I found that it happens in `drawHourLines` method, right here where the array is accessed:
```
77	hourLines[i * 4] = 0f
78	hourLines[i * 4 + 1] = top
79	hourLines[i * 4 + 2] = canvas.width.toFloat()
80	hourLines[i * 4 + 3] = top
```

Adding a few `hourLines` in the `calcHourLines` method in one of the following two lines ( + 2 instead of + 1 ) resolves the crash.
```
35	var lineCount = ((height - drawConfig.headerHeight) / config.hourHeight).toInt() + 1
36	lineCount *= (config.visibleDays + 1)
```

I just can't figure out which is the right one or if this is even the right approach...

 - Android version: 10 (shouldn't matter)
 - BetterUntis version: Confirmed on 3.4.4 from F-Droid, as well as the latest self compiled version from the develop branch

Confirmed by another person at my school, which sadly doesn't provide an anonymous login.

**Comment by SapuSeven** [comment]
I was able to reproduce this. Working on it.

**Comment by DreamyLynn** [comment]
I am not able to reproduce this. Which device are you able to reproduce it on?

**Comment by SapuSeven** [comment]
I have v3.4.5 installed on my OnePlus 6T (Android 11).
My account is not working anymore but I can still set the number of displayed days to 1 and cause a crash by randomly scrolling/zooming around.

**Comment by DreamyLynn** [comment]
Thank you. This is also affected by the school. But since it doesn't need to load a timetable, using anonymous login should work, even if it's not provided. Which school are you testing this with?

**Comment by DreamyLynn** [comment]
Found one :). Working on it.

**Comment by DreamyLynn** [comment]
I think the problem is that the `for (hour in 1 until config.hourLines.size)` loop starts at index 1, because of the next line `if (config.hourLines[hour] == config.hourLines[hour - 1]) continue`, and there is no additional element added to compensate for that.

**Comment by DreamyLynn** [comment]
I think this could be caused by the division in `calcHourLines()` in `BackgroundGridDrawer.kt`
In that case changing 
`var lineCount = ((height - drawConfig.headerHeight) / config.hourHeight).toInt() + 1` to 
`var lineCount = ((height - drawConfig.headerHeight) / config.hourHeight).roundToInt() + 1` 
should be enough. Testing it, this works in my case. Could someone confirm this?

**Comment by 3nt3** [comment]
no, still crashes for me

**Comment by SapuSeven** [comment]
Should be fixed with #271
